dev/com.ibm.ws.cdi.interfaces/src/com/ibm/wsspi/cdi/extension/WebSphereCDIExtensionMetaData.java

```diff
@@ -0,0 +1,74 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package io.openliberty.cdi.spi;
+
+import java.util.Collections;
+import java.util.Set;
+
+import javax.enterprise.inject.spi.Extension;
+
+/**
+ * This is a *marker* interface for CDI Runtime extensions. Liberty features that wish to extend CDI will need to
+ * register a service under this interface.
+ *
+ * To use this class you must implement one of the two methods. If you implement getBeanClasses() all classes returned by that
+ * method will be registered with CDI and may be used normally by application code. If you implement getExtensions() then
+ * all classes returned by that method will be treated as CDI extensions and run when creating the CDI container for each application.
+ *
+ * Classes returned from getExtensions() must implement javax.enterprise.inject.spi.Extension. They do *not* and should not be listed in a
+ * META-INF/services file. It is best practice to not put CDIExtensionMetaData and javax.enterprise.inject.spi.Extension on the same class
+ * as that would result in CDI and OSGI independently instantiating the class. Even though it is unlikely for the two instances to conflict, 
+ * it is best to keep the OSGI service and CDI extension separate. 
+ *
+ * The class that implements this interface should be annotated with @Component(service = WebSphereCDIExtensionMetaData.class). There are no
+ * other properties that can be included in the Component annotation.
+ *
+ * Here is a worked example of a complete WebSphereCDIExtensionMetaData implementation.
+ *
+ * @Component(service = WebSphereCDIExtensionMetaData.class)
+ * public class SPIMetaData implements WebSphereCDIExtensionMetaData {
+ *
+ *     @Override
+ *     public Set<Class<?>> getBeanClasses() {
+ *         Set<Class<?>> beans = new HashSet<Class<?>>();
+ *         //This will register a producer class and expose it's produced beans to applications
+ *         beans.add(ClassSPIRegisteredProducer.class);
+ *     }
+ * }
+ *
+ * A complete bundle that uses WebSphereCDIExtensionMetaData to register CDI beans can be found at:
+ * com.ibm.ws.cdi.extension_fat/test-bundles/cdi.spi.extension/
+ *
+ * Note the files com.ibm.ws.cdi.extension_fat/cdi_spi_extension.bnd and com.ibm.ws.cdi.extension_fat/publish/features/cdi.spi.extension.mf
+ * both of which are needed to include a bundle into liberty. com.ibm.ws.cdi.extension_fat/test-applications/SPIExtension.war/ contains a
+ * simple app that injects the beans provided by cdi.spi.extension.
+ */
+
+public interface CDIExtensionMetadata {
+
+    /**
+     * All classes returned by this method will be registered with CDI. All classes must be in the same archive as your WebSphereCDIExtensionMetaData.
+     */
```

Comment:
I think we need a better explanation of what "registered with CDI" means.

E.g. Are the returned classes always treated as beans, or are they only treated as beans if they have a bean defining annotation.

Would it be accurate to say "classes returned from this method will be found by CDI during type discovery so that they can then be used as beans (or interceptors etc. if annotated as such) by the application."